sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -652,6 +654,160 @@ case class StringRPad(str: Expression, len: Expression, pad: Expression)
+      'spark.apache.org'
+      > SELECT _FUNC_('http://spark.apache.org/path?query=1', 'QUERY')
+      'query=1'
+      > SELECT _FUNC_('http://spark.apache.org/path?query=1', 'QUERY', 'query')
+      '1'""")
+case class ParseUrl(children: Seq[Expression])
+  extends Expression with ImplicitCastInputTypes with CodegenFallback {
+
+  override def nullable: Boolean = true
+  override def inputTypes: Seq[DataType] = Seq.fill(children.size)(StringType)
+  override def dataType: DataType = StringType
+  override def prettyName: String = "parse_url"
+
+  // If the url is a constant, cache the URL object so that we don't need to convert url
+  // from UTF8String to String to URL for every row.
+  @transient private lazy val cachedUrl = stringExprs(0) match {
+    case Literal(url: UTF8String, _) => getUrl(url)
+    case _ => null
+  }
+
+  // If the key is a constant, cache the Pattern object so that we don't need to convert key
+  // from UTF8String to String to StringBuilder to String to Pattern for every row.
+  @transient private lazy val cachedPattern = stringExprs(2) match {
+    case Literal(key: UTF8String, _) => getPattern(key)
+    case _ => null
+  }
+
+  private lazy val stringExprs = children.toArray
+  import ParseUrl._
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    if (children.size > 3 || children.size < 2) {
+      TypeCheckResult.TypeCheckFailure(s"$prettyName function requires two or three arguments")
+    } else {
+      super[ImplicitCastInputTypes].checkInputDataTypes()
+    }
+  }
+
+  private def getPattern(key: UTF8String): Pattern = {
+    if (key != null) {
+      Pattern.compile(REGEXPREFIX + key.toString + REGEXSUBFIX)
+    } else {
+      null
+    }
+  }
+
+  private def getUrl(url: UTF8String): URL = {
+    try {
+      new URL(url.toString)
+    } catch {
+      case e: MalformedURLException => null
+    }
+  }
+
+  private def extractValueFromQuery(query: UTF8String, pattern: Pattern): UTF8String = {
+    val m = pattern.matcher(query.toString)
+    if (m.find()) {
+      UTF8String.fromString(m.group(2))
+    } else {
+      null
+    }
+  }
+
+  private def extractFromUrl(url: URL, partToExtract: UTF8String): UTF8String = {
+    if (partToExtract.equals(HOST)) {
+      UTF8String.fromString(url.getHost)
+    } else if (partToExtract.equals(PATH)) {
+      UTF8String.fromString(url.getPath)
+    } else if (partToExtract.equals(QUERY)) {
+      UTF8String.fromString(url.getQuery)
+    } else if (partToExtract.equals(REF)) {
+      UTF8String.fromString(url.getRef)
+    } else if (partToExtract.equals(PROTOCOL)) {
+      UTF8String.fromString(url.getProtocol)
+    } else if (partToExtract.equals(FILE)) {
+      UTF8String.fromString(url.getFile)
+    } else if (partToExtract.equals(AUTHORITY)) {
+      UTF8String.fromString(url.getAuthority)
+    } else if (partToExtract.equals(USERINFO)) {
+      UTF8String.fromString(url.getUserInfo)
+    } else {
+      null
```

Comment:
Since check it here is at Excutor side, it will be of no difference with current implementation. I think the point is whether we can assume that in almost all cases `part` is `Literal`.